src/systems/camera_video_recorder/CameraVideoRecorder.cc

```diff
@@ -0,0 +1,374 @@
+#include "ignition/gazebo/Conversions.hh"
+#include "ignition/gazebo/EntityComponentManager.hh"
+#include "ignition/gazebo/Events.hh"
+#include "ignition/gazebo/Util.hh"
+
+#include "CameraVideoRecorder.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+// Private data class.
+class ignition::gazebo::systems::CameraVideoRecorderPrivate
+{
+  /// \brief Callback for the video recorder service
+  public: bool OnRecordVideo(const msgs::VideoRecord &_msg,
+      msgs::Boolean &_res);
+
+  /// \brief Callback invoked in the rendering thread after a render update
+  public: void OnPostRender();
+
+  /// \brief Callback for new images
+  public: void OnImage(const msgs::Image &_msg);
+
+  /// \brief Transport node
+  public: transport::Node node;
+
+  /// \brief Mutex to protect updates
+  public: std::mutex updateMutex;
+
+  /// \brief Connection to the post-render event.
+  public: ignition::common::ConnectionPtr postRenderConn;
+
+  /// \brief Pointer to the event manager
+  public: EventManager *eventMgr = nullptr;
+
+  //// \brief Pointer to the rendering scene
+  public: rendering::ScenePtr scene;
+
+  /// \brief Pointer to the rendering camera
+  public: rendering::CameraPtr camera;
+
+  /// \brief Name of service for recording video
+  public: std::string topic;
+
+  /// \brief Camera entity.
+  public: Entity entity;
+
+  /// \brief Name of the camera
+  public: std::string cameraName;
+
+  /// \brief Image from user camera
+  public: rendering::Image cameraImage;
+
+  /// \brief Video encoder
+  public: common::VideoEncoder videoEncoder;
+
+  /// \brief Video encoding format
+  public: std::string recordVideoFormat;
+
+  /// \brief Path to save the recorded video
+  public: std::string recordVideoSavePath;
+
+  /// \brief Temporary filename of the video being encoded
+  public: std::string tmpVideoFilename;
+
+  /// \brief True to record a video from the rendering camera
+  public: bool recordVideo = false;
+
+  /// \brief Topic that the sensor publishes to
+  public: std::string sensorTopic;
+};
+
+//////////////////////////////////////////////////
+void CameraVideoRecorderPrivate::OnImage(const msgs::Image &)
+{
+  // No work is done here. We need to subscribe to the sensor to make it active.
+}
+
+//////////////////////////////////////////////////
+bool CameraVideoRecorderPrivate::OnRecordVideo(const msgs::VideoRecord &_msg,
+    msgs::Boolean &_res)
+{
+  std::lock_guard<std::mutex> lock(this->updateMutex);
+  bool record = _msg.start() && !_msg.stop();
+  this->recordVideo = record;
+
+  if (this->recordVideo)
+  {
+    this->recordVideoFormat = _msg.format();
+    this->recordVideoSavePath = _msg.save_filename();
+
+    if (this->recordVideoFormat.empty())
+      this->recordVideoFormat = "mp4";
+
+    // create filename with timestamped suffix if path is not specified
+    if (this->recordVideoSavePath.empty())
+    {
+      std::chrono::system_clock::time_point p =
```

Comment:
Can you use: `std::string str = common::systemTimeISO();`

Reply:
done. d90f78e